File: e2e/go/send_currency_tx.go
// Simple script to send a currency transaction to a local metagraph.
//
// Reads configuration from config.json and submits a currency transaction
// to the Currency L1 endpoint.
//
// Usage:
//
//	go run send_currency_tx.go
//	go run send_currency_tx.go -config other_config.json
//	go run send_currency_tx.go -generate-keypair
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"path/filepath"

	constellation "github.com/Constellation-Labs/metakit-sdk/packages/go"
)

type Config struct {
	PrivateKey    string  `json:"private_key"`
	Destination   string  `json:"destination"`
	Amount        float64 `json:"amount"`
	Fee           float64 `json:"fee"`
	CurrencyL1URL string  `json:"currency_l1_url"`
}

func loadConfig(configPath string) (*Config, error) {
	data, err := os.ReadFile(configPath)
	if err != nil {
		return nil, err
	}

	var config Config
	if err := json.Unmarshal(data, &config); err != nil {
		return nil, err
	}

	return &config, nil
}

func generateKeypairCommand() {
	keypair, err := constellation.GenerateKeyPair()
	if err != nil {
		fmt.Printf("Error generating keypair: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Generated new keypair:")
	fmt.Printf("  Private Key: %s\n", keypair.PrivateKey)
	fmt.Printf("  Public Key:  %s\n", keypair.PublicKey)
	fmt.Printf("  DAG Address: %s\n", keypair.Address)
	fmt.Println("\nSave the private key to your config.json to use it for transactions.")
}

func sendTransaction(config *Config) {
	// Validate config
	if config.PrivateKey == "" {
		fmt.Println("Error: Missing required field 'private_key' in config")
		os.Exit(1)
	}
	if config.Destination == "" {
		fmt.Println("Error: Missing required field 'destination' in config")
		os.Exit(1)
	}
	if config.CurrencyL1URL == "" {
		fmt.Println("Error: Missing required field 'currency_l1_url' in config")
		os.Exit(1)
	}

	privateKey := config.PrivateKey
	destination := config.Destination
	amount := config.Amount
	fee := config.Fee
	currencyL1URL := config.CurrencyL1URL

	// Validate private key format
	if privateKey == "YOUR_64_CHAR_HEX_PRIVATE_KEY_HERE" {
		fmt.Println("Error: Please set your private key in config.json")
		fmt.Println("Run with -generate-keypair to create a new keypair")
		os.Exit(1)
	}

	if len(privateKey) != 64 {
		fmt.Printf("Error: Private key must be 64 hex characters, got %d\n", len(privateKey))
		os.Exit(1)
	}

	// Derive address from private key
	keypair, err := constellation.KeyPairFromPrivateKey(privateKey)
	if err != nil {
		fmt.Printf("Error deriving keypair: %v\n", err)
		os.Exit(1)
	}
	sourceAddress := keypair.Address

	fmt.Printf("Source Address: %s\n", sourceAddress)
	fmt.Printf("Destination:    %s\n", destination)
	fmt.Printf("Amount:         %v tokens\n", amount)
	fmt.Printf("Fee:            %v tokens\n", fee)
	fmt.Printf("Currency L1:    %s\n", currencyL1URL)
	fmt.Println()

	// Create client
	client, err := constellation.NewCurrencyL1Client(constellation.NetworkConfig{L1URL: currencyL1URL})
	if err != nil {
		fmt.Printf("Error creating client: %v\n", err)
		os.Exit(1)
	}

	// Check health
	fmt.Println("Checking node health...")
	if !client.CheckHealth() {
		fmt.Println("Error: Currency L1 node is not responding")
		os.Exit(1)
	}
	fmt.Println("Node is healthy!")
	fmt.Println()

	// Get last reference
	fmt.Printf("Fetching last reference for %s...\n", sourceAddress)
	lastRef, err := client.GetLastReference(sourceAddress)
	if err != nil {
		fmt.Printf("Error getting last reference: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Last Reference Hash:    %s\n", lastRef.Hash)
	fmt.Printf("Last Reference Ordinal: %d\n", lastRef.Ordinal)
	fmt.Println()

	// Create transaction
	fmt.Println("Creating transaction...")
	transferParams := constellation.TransferParams{
		Destination: destination,
		Amount:      amount,
		Fee:         fee,
	}
	tx, err := constellation.CreateCurrencyTransaction(transferParams, privateKey, *lastRef)
	if err != nil {
		fmt.Printf("Error creating transaction: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("Transaction created successfully!")

	// Verify transaction locally
	fmt.Println("Verifying transaction signature...")
	result := constellation.VerifyCurrencyTransaction(tx)
	if !result.IsValid {
		fmt.Println("Error: Transaction signature verification failed!")
		os.Exit(1)
	}
	fmt.Println("Signature verified!")
	fmt.Println()

	// Submit transaction
	fmt.Println("Submitting transaction to network...")
	response, err := client.PostTransaction(tx)
	if err != nil {
		fmt.Printf("Error submitting transaction: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("Transaction submitted!")
	fmt.Printf("Transaction Hash: %s\n", response.Hash)
	fmt.Println()

	// Poll for status (optional)
	fmt.Println("Checking transaction status...")
	pending, err := client.GetPendingTransaction(response.Hash)
	if err != nil {
		fmt.Printf("Could not check status: %v\n", err)
	} else if pending != nil {
		fmt.Printf("Status: %s\n", pending.Status)
	} else {
		fmt.Println("Transaction not found in pending pool (may already be confirmed)")
	}

	fmt.Println()
	fmt.Println("Done!")
}

func main() {
	configFile := flag.String("config", "config.json", "Path to config file")
	generateKeypair := flag.Bool("generate-keypair", false, "Generate a new keypair and exit")
	flag.Parse()

	if *generateKeypair {
		generateKeypairCommand()
		return
	}

	// Get the directory of the executable/script
	execPath, err := os.Executable()
	if err != nil {
		// Fallback to working directory
		execPath, _ = os.Getwd()
	}
	scriptDir := filepath.Dir(execPath)

	// For `go run`, use the source file's directory instead
	if args := os.Args; len(args) > 0 {
		// Check if running via `go run`
		if _, err := os.Stat(filepath.Join(scriptDir, *configFile)); os.IsNotExist(err) {
			// Try current working directory
			cwd, _ := os.Getwd()
			scriptDir = cwd
		}
	}

	configPath := filepath.Join(scriptDir, *configFile)

	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		fmt.Printf("Error: Config file not found: %s\n", configPath)
		os.Exit(1)
	}

	config, err := loadConfig(configPath)
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	sendTransaction(config)
}

File: e2e/java/src/main/java/SendCurrencyTx.java
import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;
import io.constellationnetwork.metagraph.sdk.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Simple script to send a currency transaction to a local metagraph.
 *
 * Reads configuration from config.json and submits a currency transaction
 * to the Currency L1 endpoint.
 *
 * Usage:
 *   mvn exec:java
 *   mvn exec:java -Dexec.args="--config other_config.json"
 *   mvn exec:java -Dexec.args="--generate-keypair"
 */
public class SendCurrencyTx {

    private static final Gson gson = new Gson();

    static class Config {
        @SerializedName("private_key")
        String privateKey;
        String destination;
        double amount;
        double fee;
        @SerializedName("currency_l1_url")
        String currencyL1Url;
    }

    public static void main(String[] args) {
        String configFile = "../config.json";
        boolean generateKeypair = false;

        // Parse arguments
        for (int i = 0; i < args.length; i++) {
            if ("--generate-keypair".equals(args[i])) {
                generateKeypair = true;
            } else if ("--config".equals(args[i]) && i + 1 < args.length) {
                configFile = args[++i];
            } else if ("--help".equals(args[i]) || "-h".equals(args[i])) {
                printHelp();
                return;
            }
        }

        if (generateKeypair) {
            generateKeypairCommand();
            return;
        }

        // Load config
        Path configPath = Paths.get(configFile);
        if (!Files.exists(configPath)) {
            System.err.println("Error: Config file not found: " + configPath.toAbsolutePath());
            System.exit(1);
        }

        Config config;
        try {
            String content = Files.readString(configPath);
            config = gson.fromJson(content, Config.class);
        } catch (IOException e) {
            System.err.println("Error reading config: " + e.getMessage());
            System.exit(1);
            return;
        }

        sendTransaction(config);
    }

    private static void printHelp() {
        System.out.println("Usage: mvn exec:java -Dexec.args=\"[OPTIONS]\"");
        System.out.println();
        System.out.println("Options:");
        System.out.println("  --config <file>     Path to config file (default: ../config.json)");
        System.out.println("  --generate-keypair  Generate a new keypair and exit");
        System.out.println("  --help, -h          Show this help message");
    }

    private static void generateKeypairCommand() {
        Types.KeyPair keypair = Wallet.generateKeyPair();
        System.out.println("Generated new keypair:");
        System.out.println("  Private Key: " + keypair.getPrivateKey());
        System.out.println("  Public Key:  " + keypair.getPublicKey());
        System.out.println("  DAG Address: " + keypair.getAddress());
        System.out.println();
        System.out.println("Save the private key to your config.json to use it for transactions.");
    }

    private static void sendTransaction(Config config) {
        // Validate config
        if (config.privateKey == null || config.privateKey.isEmpty()) {
            System.err.println("Error: Missing required field 'private_key' in config");
            System.exit(1);
        }
        if (config.destination == null || config.destination.isEmpty()) {
            System.err.println("Error: Missing required field 'destination' in config");
            System.exit(1);
        }
        if (config.currencyL1Url == null || config.currencyL1Url.isEmpty()) {
            System.err.println("Error: Missing required field 'currency_l1_url' in config");
            System.exit(1);
        }

        String privateKey = config.privateKey;
        String destination = config.destination;
        double amount = config.amount;
        double fee = config.fee;
        String currencyL1Url = config.currencyL1Url;

        // Validate private key format
        if ("YOUR_64_CHAR_HEX_PRIVATE_KEY_HERE".equals(privateKey)) {
            System.err.println("Error: Please set your private key in config.json");
            System.err.println("Run with --generate-keypair to create a new keypair");
            System.exit(1);
        }

        if (privateKey.length() != 64) {
            System.err.println("Error: Private key must be 64 hex characters, got " + privateKey.length());
            System.exit(1);
        }

        // Derive address from private key
        Types.KeyPair keypair = Wallet.keyPairFromPrivateKey(privateKey);
        String sourceAddress = keypair.getAddress();

        System.out.println("Source Address: " + sourceAddress);
        System.out.println("Destination:    " + destination);
        System.out.println("Amount:         " + amount + " tokens");
        System.out.println("Fee:            " + fee + " tokens");
        System.out.println("Currency L1:    " + currencyL1Url);
        System.out.println();

        // Create client
        NetworkTypes.NetworkConfig networkConfig = new NetworkTypes.NetworkConfig.Builder()
                .l1Url(currencyL1Url)
                .build();
        CurrencyL1Client client = new CurrencyL1Client(networkConfig);

        // Check health
        System.out.println("Checking node health...");
        if (!client.checkHealth()) {
            System.err.println("Error: Currency L1 node is not responding");
            System.exit(1);
        }
        System.out.println("Node is healthy!");
        System.out.println();

        // Get last reference
        System.out.println("Fetching last reference for " + sourceAddress + "...");
        CurrencyTypes.TransactionReference lastRef;
        try {
            lastRef = client.getLastReference(sourceAddress);
            System.out.println("Last Reference Hash:    " + lastRef.getHash());
            System.out.println("Last Reference Ordinal: " + lastRef.getOrdinal());
        } catch (Exception e) {
            System.err.println("Error getting last reference: " + e.getMessage());
            System.exit(1);
            return;
        }
        System.out.println();

        // Create transaction
        System.out.println("Creating transaction...");
        CurrencyTypes.CurrencyTransaction tx;
        try {
            CurrencyTypes.TransferParams transferParams = new CurrencyTypes.TransferParams(
                    destination, amount, fee);
            tx = CurrencyTransaction.createCurrencyTransaction(transferParams, privateKey, lastRef);
            System.out.println("Transaction created successfully!");
        } catch (Exception e) {
            System.err.println("Error creating transaction: " + e.getMessage());
            System.exit(1);
            return;
        }

        // Verify transaction locally
        System.out.println("Verifying transaction signature...");
        Types.VerificationResult result = CurrencyTransaction.verifyCurrencyTransaction(tx);
        if (!result.isValid()) {
            System.err.println("Error: Transaction signature verification failed!");
            System.exit(1);
        }
        System.out.println("Signature verified!");
        System.out.println();

        // Submit transaction
        System.out.println("Submitting transaction to network...");
        NetworkTypes.PostTransactionResponse response;
        try {
            response = client.postTransaction(tx);
            System.out.println("Transaction submitted!");
            System.out.println("Transaction Hash: " + response.getHash());
        } catch (Exception e) {
            System.err.println("Error submitting transaction: " + e.getMessage());
            System.exit(1);
            return;
        }
        System.out.println();

        // Poll for status (optional)
        System.out.println("Checking transaction status...");
        try {
            NetworkTypes.PendingTransaction pending = client.getPendingTransaction(response.getHash());
            if (pending != null) {
                System.out.println("Status: " + pending.getStatus());
            } else {
                System.out.println("Transaction not found in pending pool (may already be confirmed)");
            }
        } catch (Exception e) {
            System.out.println("Could not check status: " + e.getMessage());
        }

        System.out.println();
        System.out.println("Done!");
    }
}
